File: test/arguments/with_range_check_code/cpp/Int4RangeCheckTest.cpp
#include "gtest/gtest.h"

#include "zserio/CppRuntimeException.h"

#include "with_range_check_code/int4_range_check/Int4RangeCheckCompound.h"

namespace with_range_check_code
{
namespace int4_range_check
{

class Int4RangeCheckTest : public ::testing::Test
{
protected:
    void checkInt4Value(int8_t value)
    {
        Int4RangeCheckCompound int4RangeCheckCompound;
        int4RangeCheckCompound.setValue(value);
        zserio::BitStreamWriter writer;
        int4RangeCheckCompound.write(writer);
        size_t writeBufferByteSize;
        const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
        zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
        const Int4RangeCheckCompound readInt4RangeCheckCompound(reader);
        ASSERT_EQ(int4RangeCheckCompound, readInt4RangeCheckCompound);
    }

    static const int8_t     INT4_LOWER_BOUND = INT8_C(-8);
    static const int8_t     INT4_UPPER_BOUND = INT8_C(7);
};

TEST_F(Int4RangeCheckTest, int4LowerBound)
{
    checkInt4Value(INT4_LOWER_BOUND);
}

TEST_F(Int4RangeCheckTest, int4UpperBound)
{
    checkInt4Value(INT4_UPPER_BOUND);
}

TEST_F(Int4RangeCheckTest, int4BelowLowerBound)
{
    try
    {
        checkInt4Value(INT4_LOWER_BOUND - 1);
        FAIL() << "Actual: no exception, Expected: zserio::CppRuntimeException";
    }
    catch (const zserio::CppRuntimeException& excpt)
    {
        ASSERT_STREQ("Value -9 of Int4RangeCheckCompound.value exceeds the range of <-8..7>!", excpt.what());
    }
}

TEST_F(Int4RangeCheckTest, int4AboveUpperBound)
{
    try
    {
        checkInt4Value(INT4_UPPER_BOUND + 1);
        FAIL() << "Actual: no exception, Expected: zserio::CppRuntimeException";
    }
    catch (const zserio::CppRuntimeException& excpt)
    {
        ASSERT_STREQ("Value 8 of Int4RangeCheckCompound.value exceeds the range of <-8..7>!", excpt.what());
    }
}

} // namespace int4_range_check
} // namespace with_range_check_code

File: test/arguments/with_range_check_code/python/Int4RangeCheckTest.py
import zserio

import WithRangeCheckCode


class Int4RangeCheckTest(WithRangeCheckCode.TestCase):
    def testInt4LowerBound(self):
        self._checkInt4Value(INT4_LOWER_BOUND)

    def testInt4UpperBound(self):
        self._checkInt4Value(INT4_UPPER_BOUND)

    def testInt4BelowLowerBound(self):
        with self.assertRaises(zserio.PythonRuntimeException):
            self._checkInt4Value(INT4_LOWER_BOUND - 1)

    def testInt4AboveUpperBound(self):
        with self.assertRaises(zserio.PythonRuntimeException):
            self._checkInt4Value(INT4_UPPER_BOUND + 1)

    def _checkInt4Value(self, value):
        int4RangeCheckCompound = self.api.Int4RangeCheckCompound(value)
        bitBuffer = zserio.serialize(int4RangeCheckCompound)
        readInt4RangeCheckCompound = zserio.deserialize(self.api.Int4RangeCheckCompound, bitBuffer)
        self.assertEqual(int4RangeCheckCompound, readInt4RangeCheckCompound)


INT4_LOWER_BOUND = -8
INT4_UPPER_BOUND = 7
